File: practice/algorithms/implementation/larrys_array/larrys_array.cpp
#include <vector>
#include <iostream>

using namespace std;

void rotate_triple(vector<int> & i_a, int i_pos)
{
    int tmp = i_a[i_pos];
    i_a[i_pos] = i_a[i_pos + 1];
    i_a[i_pos + 1] = i_a[i_pos + 2];
    i_a[i_pos + 2] = tmp;
}

void set_min_correct(vector<int> & i_a, int i_pos)
{
    int m = 1000000;
    int p = i_pos;
    for (int pos = i_pos; pos < i_a.size(); ++pos)
    {
        if (i_a[pos] < m)
        {
            m = i_a[pos];
            p = pos;
        }
    }
    
    while (p > i_pos)
    {
        if (p == (i_pos + 1))
        {
            rotate_triple(i_a, p - 1);
            p--;
        }
        else
        {
            rotate_triple(i_a, p - 2);
            rotate_triple(i_a, p - 2);
            p -= 2;
        }
    }
}

int main()
{
    int t, n;
    vector<int> a;
    
    cin >> t;
    while (t--)
    {
        cin >> n;
        a = vector<int>(n);
        for (int pos = 0; pos < n; ++pos)
        {
            cin >> a[pos];
        }
        
        for (int pos = 0; pos < n - 2; ++pos)
        {
            set_min_correct(a, pos);
        }
        
        if (a[n - 3] <= a[n - 2] && a[n - 2] <= a[n - 1])
        {
            cout << "YES" << endl;
        }
        else
        {
            cout << "NO" << endl;
        }
    }
    
    return 0;
}

File: practice/algorithms/implementation/larrys_array/larrys_array.py
#!/usr/bin/env python

def rotate_triple(i_a, i_pos):
    tmp = i_a[i_pos]
    i_a[i_pos] = i_a[i_pos + 1]
    i_a[i_pos + 1] = i_a[i_pos + 2]
    i_a[i_pos + 2] = tmp

def set_min_correct(i_arr, i_pos):
    m = min(i_arr[i_pos:])
    p = i_arr.index(m)
    while p > i_pos:
        if p == i_pos + 1:
            rotate_triple(i_arr, p - 1)
            p -= 1
        else:
            rotate_triple(i_arr, p - 2)
            rotate_triple(i_arr, p - 2)
            p -= 2

t = int(raw_input().strip())
for _ in range(t):
    n = int(raw_input().strip())
    a = map(int, raw_input().strip().split())
    for POS in range(n - 2):
        set_min_correct(a, POS)
    b = a[n - 3 :]
    if b == sorted(b):
        print "YES"
    else:
        print "NO"

